pkg/reconciler/v1alpha1/certificate/certificate.go

```diff
@@ -0,0 +1,223 @@
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+	} else if _, err := c.updateStatus(knCert); err != nil {
+		logger.Warnw("Failed to update certificate status", zap.Error(err))
+		c.Recorder.Eventf(knCert, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update status for Certificate %s: %v", key, err)
+		return err
+	}
+	if err != nil {
+		c.Recorder.Event(knCert, corev1.EventTypeWarning, "InternalError", err.Error())
+	}
+	return err
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, knCert *v1alpha1.Certificate) error {
+	logger := logging.FromContext(ctx)
+
+	knCert.SetDefaults(ctx)
+	knCert.Status.InitializeConditions()
+
+	logger.Info("Reconciling Cert-Manager certificate for Knative cert %s/%s.", knCert.Namespace, knCert.Name)
+	cmConfig := config.FromContext(ctx).CertManager
+	cmCert := resources.MakeCertManagerCertificate(cmConfig, knCert)
+	cmCert, err := c.reconcileCMCertificate(ctx, knCert, cmCert)
+	if err != nil {
+		return err
+	}
+	setStatus(knCert, cmCert)
+	return nil
+}
+
+func (c *Reconciler) reconcileCMCertificate(ctx context.Context, knCert *v1alpha1.Certificate, desired *cmv1alpha1.Certificate) (*cmv1alpha1.Certificate, error) {
+	logger := logging.FromContext(ctx)
+	cmCert, err := c.cmCertificateLister.Certificates(desired.Namespace).Get(desired.Name)
+	if apierrs.IsNotFound(err) {
+		cmCert, err = c.certManagerClient.CertmanagerV1alpha1().Certificates(desired.Namespace).Create(desired)
+		if err != nil {
+			logger.Errorw("Failed to create Cert-Manager certificate", zap.Error(err))
+			c.Recorder.Eventf(knCert, corev1.EventTypeWarning, "CreationFailed",
+				"Failed to create Cert-Manager Certificate %s/%s: %v", desired.Name, desired.Namespace, err)
+			return nil, err
+		}
+		c.Recorder.Eventf(knCert, corev1.EventTypeNormal, "Created",
+			"Created Cert-Manager Certificate %s/%s", desired.Namespace, desired.Name)
+	} else if err != nil {
+		return nil, err
+	} else if !equality.Semantic.DeepEqual(cmCert.Spec, desired.Spec) {
+		copy := cmCert.DeepCopy()
+		copy.Spec = desired.Spec
+		updated, err := c.certManagerClient.CertmanagerV1alpha1().Certificates(copy.Namespace).Update(copy)
+		if err != nil {
+			logger.Errorw("Failed to update Cert-Manager Certificate", zap.Error(err))
+			c.Recorder.Eventf(knCert, corev1.EventTypeWarning, "UpdateFailed",
+				"Failed to create Cert-Manager Certificate %s/%s: %v", desired.Namespace, desired.Name, err)
+			return nil, err
+		}
+		c.Recorder.Eventf(knCert, corev1.EventTypeNormal, "Updated",
+			"Updated Spec for Cert-Manager Certificate %s/%s", desired.Namespace, desired.Name)
+		return updated, nil
+	}
+	return cmCert, nil
+}
+
+func setStatus(knCert *v1alpha1.Certificate, cmCert *cmv1alpha1.Certificate) {
```

Comment:
I think it might be clearer to just inline this, the method name doesn't help me and this isn't very long.